### PR TITLE
Implement Revision for Roaring

### PR DESCRIPTION
**Pull request by emmanuel-keller**
A benchmark is validating the best serialisation for both space and time

```
Bincode::default, Bincode::options, Direct, Ratio direct/bincode::options
Elapsed - 50 > 30 > 26 - 0.8666667
Size: 8228 > 8223 > 8220  - 0.99963516
```

Tests passing:

```
cargo test
    Finished test [unoptimized + debuginfo] target(s) in 0.04s
     Running unittests src/lib.rs (target/debug/deps/revision-565bf45e49921eb4)

running 45 tests
test implementations::bound::tests::test_bound_excluded ... ok
test implementations::bound::tests::test_bound_unbounded ... ok
test implementations::bound::tests::test_bound_included ... ok
test implementations::cow::test::cow_borrow ... ok
test implementations::chrono::tests::test_datetime_min ... ok
test implementations::chrono::tests::test_datetime_max ... ok
test implementations::boxes::tests::test_box ... ok
test implementations::cow::test::cow_owned ... ok
test implementations::decimal::tests::test_decimal_min ... ok
test implementations::decimal::tests::test_decimal_max ... ok
test implementations::duration::tests::test_string ... ok
test implementations::option::tests::test_option_none ... ok
test implementations::option::tests::test_option_some ... ok
test implementations::primitives::tests::test_bool ... ok
test implementations::primitives::tests::test_char ... ok
test implementations::primitives::tests::test_f32 ... ok
test implementations::primitives::tests::test_f64 ... ok
test implementations::primitives::tests::test_i128 ... ok
test implementations::primitives::tests::test_i16 ... ok
test implementations::primitives::tests::test_i32 ... ok
test implementations::primitives::tests::test_i64 ... ok
test implementations::primitives::tests::test_i8 ... ok
test implementations::primitives::tests::test_isize ... ok
test implementations::primitives::tests::test_u128 ... ok
test implementations::primitives::tests::test_u16 ... ok
test implementations::primitives::tests::test_u32 ... ok
test implementations::primitives::tests::test_u64 ... ok
test implementations::primitives::tests::test_u8 ... ok
test implementations::primitives::tests::test_usize ... ok
test implementations::result::tests::test_result_err ... ok
test implementations::result::tests::test_result_ok ... ok
test implementations::roaring::tests::test_roaring_bitmap ... ok
test implementations::roaring::tests::test_roaring_treemap ... ok
test implementations::string::tests::test_string ... ok
test implementations::trees::tests::test_btreemap ... ok
test implementations::tuple::tests::test_tuple_2 ... ok
test implementations::tuple::tests::test_tuple_3 ... ok
test implementations::trees::tests::test_hashmap ... ok
test implementations::tuple::tests::test_tuple_4 ... ok
test implementations::uuid::tests::test_uuid ... ok
test implementations::vecs::tests::test_vec ... ok
test implementations::tuple::tests::test_tuple_5 ... ok
test implementations::wrapping::tests::test_wrapping ... ok
test implementations::regex::tests::test_regex ... ok
test implementations::roaring::tests::test_roaring_serialization_benchmark ... ok

test result: ok. 45 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/test.rs (target/debug/deps/test-ebd421389299b9f1)

running 1 test
test test_enum ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests revision

running 1 test
test src/lib.rs - Revisioned (line 24) ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.13s
```

Format passed:

```
cargo fmt --all -- --check
```

### CODE REVIEW

File: src/implementations/roaring.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
+use super::super::Error;
+use super::super::Revisioned;
+use roaring::{RoaringBitmap, RoaringTreemap};
+
+impl Revisioned for RoaringTreemap {
+	#[inline]
+	fn serialize_revisioned<W: std::io::Write>(&self, writer: &mut W) -> Result<(), Error> {
+		bincode::serialize_into(writer, self)
```
Comment (tobiemh) [inline src/implementations/roaring.rs]: Do we benefit from using custom bincode options here @emmanuel-keller?

```rs
		bincode::options()
			.with_no_limit()
			.with_little_endian()
			.with_varint_encoding()
			.reject_trailing_bytes()
			.serialize_into(writer, self)
```

File: src/implementations/roaring.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+			.map_err(|ref err| Error::Serialize(format!("{:?}", err)))
+	}
+
+	#[inline]
+	fn deserialize_revisioned<R: std::io::Read>(reader: &mut R) -> Result<Self, Error> {
+		bincode::deserialize_from(reader)
```
Comment (tobiemh) [inline src/implementations/roaring.rs]: bincode::options()
			.with_no_limit()
			.with_little_endian()
			.with_varint_encoding()
			.reject_trailing_bytes()
			.deserialize_from(reader)

File: src/implementations/roaring.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+	fn revision() -> u16 {
+		1
+	}
+}
+
+impl Revisioned for RoaringBitmap {
+	#[inline]
+	fn serialize_revisioned<W: std::io::Write>(&self, writer: &mut W) -> Result<(), Error> {
+		bincode::serialize_into(writer, self)
```
Comment (tobiemh) [inline src/implementations/roaring.rs]: Same as above?

File: src/implementations/roaring.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+			.map_err(|ref err| Error::Serialize(format!("{:?}", err)))
+	}
+
+	#[inline]
+	fn deserialize_revisioned<R: std::io::Read>(reader: &mut R) -> Result<Self, Error> {
+		bincode::deserialize_from(reader)
```
Comment (tobiemh) [inline src/implementations/roaring.rs]: Same as above?

File: src/implementations/roaring.rs (hunk 2 above)
Comment (emmanuel-keller) [reply src/implementations/roaring.rs]: A benchmark evaluating the best serialization option has been added

**Review by tobiemh — APPROVED** [review]

**Pull request merged**